File: src/jl/analysis/interval_arithmetic.jl

# Abstract Interpretation
# Cousot, Patrick; Cousot, Radhia (1976). "Static determination of dynamic properties of programs"

# supports following elementary functions:
# +, *, -, /
# exp, log, sqrt, ^{float}, min, max

struct Interval
    low::Real
    high::Real
end
Base.:(==)(x::Interval, y::Interval) = x.low == y.low && x.high == y.high

function Interval(x::Real)
    return Interval(x, x)
end

Base.:∪(x::Interval, y::Interval) = Interval(min(x.low, y.low), max(x.high, y.high)) # over-approximate if disjoint

Base.:+(x::Interval, y::Interval) = Interval(x.low + y.low, x.high + y.high)
Base.:-(x::Interval, y::Interval) = Interval(x.low - y.high, x.high - y.low)
function Base.:*(x::Interval, y::Interval)
    E = [x.low * y.low, x.low * y.high, x.high * y.low, x.high * y.high]
    E = E[.!isnan.(E)]
    return Interval(minimum(E), maximum(E))
end
function Base.:/(x::Interval, y::Interval)
    if y.low != 0 && y.high != 0
        y = Interval(1/y.high, 1/y.low)
    elseif y.low != 0
        y = Interval(-Inf, 1/y.low)
    elseif y.high != 0
        y = Interval(1/y.high, Inf)
    else
        error("Division by zero.")
    end
    return x * y
end

Base.exp(x::Interval) = Interval(exp(x.low), exp(x.high))
Base.log(x::Interval) = Interval(log(x.low), log(x.high))
Base.sqrt(x::Interval) = Interval(sqrt(x.low), sqrt(x.high))
function Base.:^(x::Interval, n::Real)
    if isodd(n)
        return Interval(x.low^n, x.high^n)
    else
        if x.low >= 0
            return Interval(x.low^n, x.high^n)
        else
            return Interval(0., x.high^n)
        end
    end
end
function Base.:^(x::Interval, n::Interval)
    if n.low == n.high
        return x^n.low
    else
        # fallback
        return Interval(0., Inf)
    end
end
Base.min(x::Interval, y::Interval) = Interval(min(x.low, y.low), min(x.high, y.high))
Base.max(x::Interval, y::Interval) = Interval(max(x.low, y.low), max(x.high, y.high))

logistic(x::Interval) = Interval(0., 1.)

Base.clamp(x::Interval, a::Interval, b::Interval) = max(a, min(x, b))

const SYMBOL_TO_FUNC = Dict(
    :+ => +,
    :* => *,
    :- => -,
    :/ => /,
    :exp => exp,
    :log => log,
    :sqrt => sqrt,
    :^ => ^,
    :min => min,
    :max => max,
    :logistic => logistic,
    :clamp => clamp,
)


mutable struct StaticIntervalEvaluator <: NodeVisitor
    valuation::Dict{Symbol, Interval}
end

function visit(visitor::StaticIntervalEvaluator, node::SyntaxNode)::Interval
    @assert kind(node) != K"="
    if kind(node) == K"Integer" || kind(node) == K"Float"
        # Build intervals from constants.
        return Interval(node.val)
    end

    if kind(node) == K"Identifier"
        if !haskey(visitor.valuation, node.val)
            if node.val == :I
                return Interval(0, 1) # Identity matrix form LinearAlgebra package
            end
            # no valuation for identifier found -> overapproximate as arbitary real number
            @warn "Unknown symbol $(node.val) encountered."
            return Interval(-Inf, Inf)
        end
        # return current interval valuation for identifier
        return visitor.valuation[node.val]
    end

    if kind(node) == K"ref"
        # array elements are all masked as one interval
        return visit(visitor, node[1])
    end

    if kind(node) == K"call" || kind(node) == K"dotcall"
        # map call to interval arithmetic operation

        # get call name node and remove from children
        children = copy(node.children)
        if JuliaSyntax.is_infix_op_call(node)
            call_name_node = children[2]
            deleteat!(children, 2)
        elseif JuliaSyntax.is_postfix_op_call(node)
            call_name_node = children[end]
            deleteat!(children, length(children))
        else
            call_name_node = children[1]
            deleteat!(children, 1)
        end

        # map call name node to function name (Symbol)
        if kind(call_name_node) in (K"+", K"-", K"*", K"/", K"^")
            # special operations
            fname = Symbol(string(kind(call_name_node)))
        elseif kind(call_name_node) == K"Identifier"
            # arbitary identifier (e.g. min, max, exp, ...)
            fname = call_name_node.val
        else
            # fallback
            fname = sourcetext(call_name_node)
        end

        if haskey(visitor.valuation, fname)
            # function can also be masked
            return visitor.valuation[fname]
        end

        # check that we can map to function
        if !haskey(SYMBOL_TO_FUNC, fname)
            # if not overapproximate as arbitary real number
            @warn "Unsupported function $fname."
            return Interval(-Inf, Inf)
        end

        # evaluate interval operation
        func = SYMBOL_TO_FUNC[fname]
        args = [visit(visitor, child) for child in children]
        return func(args...)
    end

    if kind(node) == K"parens"
        # handle parenthesis
        return visit(visitor, node[1])
    end

    if kind(node) == K"macrocall" && node[1].val == Symbol("@.")
        return visit(visitor, node[2])
    end

    if kind(node) == K"vect" && length(node.children) > 0
        # arrays are approximated by having one interval for *all* elements
        args = [visit(visitor, child) for child in node.children]
        return reduce(∪, args)
    end

    error("Encountered unsupported node $(node) with kind $(kind(node))")
end

function _static_interval_eval(cfg_progr_repr::CFGRepresentation, node_to_evaluate::SyntaxNode, valuation::Dict{Symbol, Interval})
    identifiers = get_identifiers_read_in_syntaxnode(cfg_progr_repr.scoped_tree, node_to_evaluate)
    _, cfgnode = get_cfgnode_for_syntaxnode(cfg_progr_repr, node_to_evaluate)
    
    # update valuation by trying to estimate interval for each identifier read in syntaxnode
    for identifier in identifiers
        identifier_name = identifier.val
        if !haskey(valuation, identifier_name)

            rds = get_RDs(cfg_progr_repr.scoped_tree, cfgnode, identifier)
            if length(rds) == 0
                # no reaching definition found
            else # length(rds) > 0
                intervals = Interval[]
                for rd in rds
                    if rd.type == ASSIGN_NODE
                        @assert kind(rd.syntaxnode) == K"=" "Unsupported assign node $(rd.syntaxnode)"
                        rhs = rd.syntaxnode[2]
                        push!(intervals, _static_interval_eval(cfg_progr_repr, rhs, valuation))
                    else # else rd.type == FUNCARG_NODE, we do not evaluate
                        push!(intervals, Interval(-Inf, Inf))
                    end
                end
                valuation[identifier_name] = reduce(∪, intervals)
            end
        end
    end
    return visit(StaticIntervalEvaluator(valuation), node_to_evaluate)
end

# Assumptions: SSA and only elementary functions -> all definitions in same scope
# Only exception: one assignment per if branch
# Elements of array all have same interval
function static_interval_eval(cfg_progr_repr::CFGRepresentation, node_to_evaluate::SyntaxNode, valuation::Dict{Symbol, Interval})::Interval
    try
        return _static_interval_eval(cfg_progr_repr, node_to_evaluate, valuation)
    catch e
        if e isa StackOverflowError
            # we could catch recusion already in _static_interval_eval, but this is simpler
            return Interval(-Inf, Inf)
        else
            rethrow(e)
        end
    end
end

File: src/py/analysis/interval_arithmetic.py
import math
import typing
from collections import deque
from ast_utils.utils import get_call_name
from ast_utils.scoped_tree import ScopedTree, Assignment, FunctionDefinition
from analysis.data_control_flow import *
from copy import copy

class Interval:
    def __init__(self, low: float, high: float = None) -> None:
        self.low = low        
        self.high = high if high is not None else low
    def __repr__(self):
        return f"[{self.low}, {self.high}]"
    def __eq__(self, __value: object) -> bool:
        if not isinstance(__value, Interval):
            return False
        else:
            return self.low == __value.low and self.high == __value.high

def add(x: Interval, y: Interval) -> Interval:
    return Interval(x.low + y.low, x.high + y.high)

def sub(x: Interval, y: Interval) -> Interval:
    return Interval(x.low - y.high, x.high - y.low)

def usub(x: Interval) -> Interval:
    return Interval(-x.high, -x.low)

def mul(x: Interval, y: Interval) -> Interval:
    E = [x.low * y.low, x.low * y.high, x.high * y.low, x.high * y.high]
    E = [e for e in E if not math.isnan(e)]
    return Interval(min(E), max(E))

def div(x: Interval, y: Interval) -> Interval:
    if y.low == -math.inf and y.high == math.inf:
        # 1/y.high, 1/y.low == 0, 0
        return Interval(-math.inf, math.inf)
    
    if y.low != 0 and y.high != 0:
        y = Interval(1/y.high, 1/y.low)
    elif y.low != 0:
        y = Interval(-math.inf, 1/y.low)
    elif y.high != 0:
        y = Interval(1/y.high, math.inf)
    else:
        raise ValueError("Division by zero.")
    
    return mul(x, y)

def pow(x: Interval, y: Interval) -> Interval:
    if isinstance(y, Interval):
        if y.low == y.high:
            n = y.low
        else:
            return Interval(0, math.inf)
    n == y

    if n % 2 != 0:
        return Interval(x.low ** n, x.high ** n)
    else:
        if x.low >= 0:
            return Interval(x.low ** n, x.high ** n)
        else:
            return Interval(0., x.high ** n)


def minimum(x: Interval, y: Interval) -> Interval:
    return Interval(min(x.low, y.low), min(x.high, y.high))

def maximum(x: Interval, y: Interval) -> Interval:
    return Interval(max(x.low, y.low), max(x.high, y.high))

def sqrt(x: Interval) -> Interval:
    return Interval(math.sqrt(x.low), math.sqrt(x.high))

def log(x: Interval) -> Interval:
    low = math.log(x.low) if x.low != 0 else -math.inf
    high = math.log(x.high) if x.high != 0 else -math.inf

    return Interval(low, high)

def exp(x: Interval) -> Interval:
    return Interval(math.exp(x.low), math.exp(x.high))

def union(x: Interval, y: Interval) -> Interval:
    return Interval(min(x.low, y.low), max(x.high, y.high)) # over-approximate if disjoint
    
def ifelse(test: Interval, x: Interval, y: Interval) -> Interval:
    return union(x, y)

def invlogit(x: Interval) -> Interval:
    return Interval(0, 1)

def eq(x: Interval, y: Interval) -> Interval:
    return Interval(0, 1)

def no_op(x: Interval, *args) -> Interval:
    return x

def clip(x: Interval, a: Interval, b:Interval) -> Interval:
    return Interval(a.low, b.high)

def erf(x: Interval) -> Interval:
    return Interval(-1, 1)

def ones(*args) -> Interval:
    return Interval(1.,1.)

def prod(x: Interval) -> Interval:
    # x is array
    if 0 <= x.low and x.high <= 1:
        return Interval(0.,1.) # product of aribtrary many 0<=x<=1 is in [0,1]
    return Interval(-math.inf,math.inf)

# Identity Matrix
def eye(*args) -> Interval:
    return Interval(0.,1.)

import ast

SYMBOL_TO_FUNC = {
    ast.Add: add,
    ast.Sub: sub,
    ast.USub: usub,
    ast.Mult: mul,
    ast.Div: div,
    ast.Pow: pow,
    'sqrt': sqrt,
    'exp': exp,
    'log': log,
    'maximum': maximum,
    'minimum': minimum,

    # pytensor
    'ifelse': ifelse,
    'switch': ifelse,
    'invlogit': invlogit,
    'outer': mul, # outer product
    'eq': eq,
    'flatten': no_op,
    'stack': no_op,
    'reshape': no_op,
    'repeat': no_op,
    'clip': clip,
    'erf': erf,
    'ones': ones,
    'prod': prod,
    'constant': no_op, # pm.math.constant
    'eye': eye,
}

from functools import reduce

class StaticIntervalEvaluator(ast.NodeVisitor):
    def __init__(self, valuation: dict[str, Interval]):
        super().__init__()
        self.valuation = valuation

    def visit(self, node: ast.AST) -> Interval:
        if isinstance(node, (ast.Constant, ast.List, ast.Name, ast.Subscript, ast.UnaryOp, ast.BinOp, ast.Call, ast.Attribute, ast.Return)):
            return super().visit(node)
        if isinstance(node, ast.Expr):
            return self.visit(node.value)
        print(f"Encountered unsupported node {node}: {ast.unparse(node)}")
        return Interval(-math.inf, math.inf)
        
    def visit_Constant(self, node: ast.Constant) -> Interval:
        # Build intervals from constants.
        return Interval(node.value)
    
    def visit_List(self, node: ast.List) -> Interval:
        if len(node.elts) == 0:
            return Interval(-math.inf, math.inf)
        else:
            # one interval for all elements
            return reduce(union, [self.visit(arg) for arg in node.elts])
    
    def visit_Name(self, node: ast.Name) -> Interval:
        if node.id in self.valuation:
            # return current interval valuation for identifier
            return self.valuation[node.id]
        else:
            # no valuation for identifier found -> overapproximate as arbitary real number
            print(f"Warning: Unknown symbol {node.id} encountered.")
            return Interval(-math.inf, math.inf)
        
    def visit_Subscript(self, node: ast.Subscript) -> Interval:
        # # array elements are all masked as one interval
        return self.visit(node.value)
    
    def visit_UnaryOp(self, node: ast.UnaryOp) -> Interval:
        operand = self.visit(node.operand)
        func = SYMBOL_TO_FUNC[type(node.op)]
        return func(operand)

    
    def visit_BinOp(self, node: ast.BinOp) -> Interval:
        left = self.visit(node.left)
        right = self.visit(node.right)
        func = SYMBOL_TO_FUNC[type(node.op)]
        return func(left, right)
    
    def visit_Call(self, node: ast.Call) -> Interval:
        # map call to interval arithmetic operation

        call_name = get_call_name(node)
        if call_name in self.valuation:
            # functions can also be masked
            return self.valuation[call_name]

        if call_name in SYMBOL_TO_FUNC:
            func = SYMBOL_TO_FUNC[call_name]

            values = [self.visit(arg) for arg in node.args]
            if (len(node.args) == 0 or
                (call_name in ('maximum', 'minimum') and len(node.args) == 1)):
                # support x.exp(), etc
                assert isinstance(node.func, ast.Attribute)
                values.insert(0, self.visit(node.func.value))

            # # evaluate interval operation
            return func(*values)
        elif call_name == 'array' and isinstance(node.args[0], ast.List):
            # np array literal
            list = node.args[0]
            return self.visit(list)
        else:
            print(f"Unsupported function {call_name}.")
            # overapproximate as arbitary real number
            return Interval(-math.inf, math.inf)  
        
    def visit_Attribute(self, node: ast.Attribute) -> Interval:
        if node.attr == 'shape':
            # np shape
            return Interval(0, math.inf)
        elif node.attr == 'T':
            # transpose
            return self.visit(node.value)
        else:
            print(f"Warning: Unknown attribute {node.attr} encountered.")
            return Interval(-math.inf, math.inf)
        
    def visit_Return(self, node: ast.Return) -> Interval:
        return self.visit(node.value)


def _static_interval_eval(scoped_tree: ScopedTree, node_to_evaluate: ast.AST, valuation: typing.Dict[str, Interval]):
    identifiers = get_identifiers_read_in_syntaxnode(scoped_tree, node_to_evaluate)
    _, cfgnode = scoped_tree.get_cfgnode_for_syntaxnode(node_to_evaluate)
    
    # update valuation by trying to estimate interval for each identifier read in syntaxnode
    for identifier in identifiers:
        identifier_name = identifier.id
        if identifier_name not in valuation:
            is_function, function = maybe_get_user_function(scoped_tree, identifier)
            if is_function:
                # we estimate the interval of return expression for *any* input
                cfg = scoped_tree.get_cfg_for_function_syntaxnode(function.node)
                intervals = [
                    _static_interval_eval(scoped_tree, get_return_expr(cfgnode), valuation)
                    for cfgnode in cfg.nodes if isinstance(cfgnode, ReturnNode)
                ]
                valuation[identifier_name] = reduce(union, intervals)
            else:
                rds = get_RDs(scoped_tree, cfgnode, identifier)
                if len(rds) > 0:
                    intervals = []
                    for rd in rds:
                        if isinstance(rd, AssignNode):
                            assert isinstance(rd.syntaxnode, ast.Assign), f"Unsupported assign node {rd.syntaxnode}"
                            rhs = rd.syntaxnode.value
                            intervals.append(_static_interval_eval(scoped_tree, rhs, valuation))
                        else: # we do not evaluate
                            intervals.append(Interval(-math.inf, math.inf))
                    valuation[identifier_name] = reduce(union, intervals)

    return StaticIntervalEvaluator(valuation).visit(node_to_evaluate)


# Assumptions: SSA and only elementary functions -> all definitions in same scope
# Only exception: one assignment per if branch
# Elements of array all have same interval
def static_interval_eval(scoped_tree: ScopedTree, node_to_evaluate: ast.AST, valuation: typing.Dict[str, Interval]) ->Interval:
    try:
        return _static_interval_eval(scoped_tree, node_to_evaluate, valuation)
    except RecursionError as e:
        # we could catch recusion already in _static_interval_eval, but this is simpler
        return Interval(-math.inf, math.inf)
